### PR TITLE
Experimental

### PR DESCRIPTION
**Pull request by z3t0**
Added Sharp Decoding Support.
Please test and report results if you want this to become a feature and want this plugin to continue in its development.

### CODE REVIEW

**Comment by z3t0** [comment]
I added 2 more commits, fixing some syntax mistakes

**Comment by z3t0** [comment]
I have verified that this is includes all the changes in the #83 so please test and report results.

File: changelog.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Change Log
+All notable changes to this project will be documented in this file.
+This project adheres to [Semantic Versioning](http://semver.org/).
+
+## [1.0.1] - 2015-03-08
+### Added
+- Support for Sharp Decoding
```
Comment (crash7) [inline changelog.md]: It would be nice to add the Aiwa support in this file too!

File: IRremote.h [hunk 2]
```diff
@@ -53,6 +53,9 @@ class decode_results {
 #define SAMSUNG 11
 #define LG 12
 #define WHYNTER 13
+#define Sharp 14
+#define AIWA_RC_T501 15
+>>>>>>> crash7-master
```
Comment (crash7) [inline IRremote.h]: Merge conflicts line, remove this!

File: IRremote.cpp [hunk 3]
```diff
@@ -618,6 +631,7 @@ long IRrecv::decodeSony(decode_results *results) {
   return DECODED;
 }
 
+<<<<<<< HEAD
```
Comment (crash7) [inline IRremote.cpp]: Merge conflicts/diff line, remove this!

**Comment by z3t0** [comment]
@crash7 Thank you very much for fixing up my merge mistakes. I merged it while half asleep and thought I had everything fixed properly. I'll make those changes and commit.

**Comment by z3t0** [comment]
Just pushed. Please test and report results.

**Comment by z3t0** [comment]
@TheMabb Reported this:
```
Arduino: 1.6.1 (Windows 8.1), Board: "Arduino Uno"

C:\Program Files (x86)\Arduino\libraries\IRremote\IRremote.cpp: In member function 'long int IRrecv::decodeWhynter(decode_results*)':

C:\Program Files (x86)\Arduino\libraries\IRremote\IRremote.cpp:753:51: error: a function-definition is not allowed here before '{' token

long IRrecv::decodeSharp(decode_results *results) {

                                               ^
C:\Program Files (x86)\Arduino\libraries\IRremote\IRremote.cpp:1357:1: error: expected '}' at end of input

}

^

C:\Program Files (x86)\Arduino\libraries\IRremote\IRremote.cpp:1357:1: error: expected '}' at end of input

Error compiling.

This report would have more information with
"Show verbose output during compilation"
enabled in File > Preferences.
```

I am trying to find the problem. But not much luck yet.

**Comment by crash7** [comment]
Hi, just compiled all the examples with Arduino 1.6.1 (Windows 7 x64) without any errors.
Maybe he downloaded a broken version?

**Comment by z3t0** [comment]
@crash7 I don't think so, he said that he compiled the beta release (which is this branch) and I have tested it to get the same error. Please test this [beta])https://github.com/shirriff/Arduino-IRremote/releases/tag/BETA) release here.

**Comment by z3t0** [comment]
I'm sure the error is just a missing '{' or something like that but I don't have an IDE installed at the moment to check with.

**Comment by joshuajnoble** [comment]
I'm trying to understand what all is going on here, is this just for adding Sharp decode? Is there a less noisy way to do this perhaps?

**Comment by crash7** [comment]
@z3t0 yeah, you were right:
```
long IRrecv::decodeWhynter(decode_results *results) {
....
// data bits
  for (int i = 0; i < WHYNTER_BITS; i++) {
    if (!MATCH_MARK(results->rawbuf[offset], WHYNTER_BIT_MARK)) {
      return ERR;
    }
    offset++;
    if (MATCH_SPACE(results->rawbuf[offset], WHYNTER_ONE_SPACE)) {
      data = (data << 1) | 1;
    } 
    else if (MATCH_SPACE(results->rawbuf[offset],WHYNTER_ZERO_SPACE)) {
 
  // trailing mark
  if (!MATCH_MARK(results->rawbuf[offset], WHYNTER_BIT_MARK)) {
    return ERR;
  }
  // Success
  results->bits = WHYNTER_BITS;
  results->value = data;
  results->decode_type = WHYNTER;
  return DECODED;
}
```
Missing ```if``` and ```for``` log closing braces.

Seems that a previous merge broke this [file](https://github.com/shirriff/Arduino-IRremote/commits/experimental/IRremote.cpp), check the implementation [here](https://github.com/shirriff/Arduino-IRremote/blob/e04e7d31b7d8269bf4f8e4201ee9cf3b0c36561f/IRremote.cpp#L688).

If you like i can submit a patch to this branch.

**Comment by z3t0** [comment]
@crash7 I would really appreciate if you could submit a patch because I use Vim for most of my editing and it's in a weird state right now and I am working on getting Syntax Checking for C/C++ files. I don't want to submit a patch which could possibly have mistakes.

@joshuajnoble This pull request is to merge new features into the master branch. It is basically where all the PRs get merged into and tested. Once we get confirmation results we will merge it into the master branch and make a new release. I was originally planning to make this specific PR for the Sharp Decoding support but I did not do so since no one responded with their results. That is, until I know that these changes are completely stable I cannot merge it into the master for fear or releasing an unstable build. 

It would be great if you could help us by testing and reporting back results, because I don't have either Sharp or Aiwa Equipment to test this code with.

**Comment by joshuajnoble** [comment]
> This pull request is to merge new features into the master branch.

Adding multiple features to a single PRs doesn't as well as a discrete set of functionality that can be tested and then integrated into the branch. If there's multiple new features being merged here then multiple PRs should be created for each of them, otherwise accurately understanding what's being merged and being able to roll it back in the event that bugs are discovered later is much harder.

**Comment by z3t0** [comment]
@joshuajnoble But isn't it essentially the same because we do not add features in this PR specifically. We merge existing PRs and test/patch them. Please understand that I would love any feedback and am not trying to discredit you, rather this is a discussion and I would like to benefit from it so that I can better help out with this project.

**Comment by joshuajnoble** [comment]
Sure, and there's inherently nothing wrong with it, I just personally find this a very noisy way to do it that makes it difficult to know what's actually being proposed for the library. For instance, I have no Sharp devices or Aiwa devices, which is why I haven't done anything with them. However #110 can be tested and merged right now. Adding all of those together to do a much larger merge in the future doesn't seem to make as much sense to me, unless this is really just a giant PR that will never actually be merged?

**Comment by z3t0** [comment]
@joshuajnoble That makes perfect sense and is definitely better than what's going on in this branch. Also, it definitely means faster releases to the general public. I don't know why but it just made more sense to me to have an experimental branch... Maybe because some of the PRs are quite old and it would be better to create a ready-made testing zip file to get more people to try it. When I started sifting through the PRs I noticed many of them were way out of sync, so I initially thought this would be a good approach. Also, not to be offensive towards anyone, but when I merged the PR that added Sharp support into here I thought that it would be tested very soon because the person that had created the PR seemed quite active... but unfortunately I still have no way to know if it works. 

Theoretically this is supposed to be a test bench before merging into the master branch but practically it seems as you suggested that things will never get merged because testing becomes a problem. 

Something like #110 as you mentioned can be directly added to master, but what I was intending to do is do the 'development' phase inside experimental in an attempt to ensure that the master branch is always stable, consequently every notable merge into Master will be a new version which will be released.

Maybe after this we should just use the master branch? But my idea was to keep that branch as stable as possible. Let me know what you think

**Comment by crash7** [comment]
@z3t0 @joshuajnoble sorry to interrupt.

I suggest the following:
1. Let's split this giant PR in different PRs
2. Let's keep the master branch as the stable one
3. Only merge PRs that are tested by at least one of the main colaborators (we could also use labels like "NEED TESTERS") and don't break existing functionality.
4. After a merge into master, new release

If we follow this (especially 3), the master branch would be "always" stable.
As @joshuajnoble it would be easier to test & merge small PRs.

**Comment by z3t0** [comment]
@crash7 that sounds like an awesome idea, the only problem I can think of is that many of the PRs are extremely out of sync. Should we just try and sync them and patch up merge errors or recreate the PRs?

**Comment by joshuajnoble** [comment]
I vote for syncing old PRs. With a library like this it's usually very easy as the added functionality tends towards the additive rather than the transformative. If someone is proposing rewriting core functionality then that needs to be its own PR that can be tested independent of a device.

**Comment by z3t0** [comment]
@joshuajnoble Agreed. So as for splitting this into separate parts we should probably just split up the workload? I'm kinda distracted with some other things but if you let me know which PRs to work on I can get them up to date soon.

**Comment by crash7** [comment]
@z3t0 the first aproach should be adding a comment to the PRs asking the owner to update it (and label it with something like "NEED UPDATE"). If the owner doesn't appears and we can test it or make sure that doesn't break anything, let's update/recreate the PRs and label them for review.

The critical problem would be updating PRs that are too old and the current stable codebase changed dramatically (probably not this case).

**Comment by z3t0** [comment]
@crash7 OK I'll get onto it.

**Comment by joshuajnoble** [comment]
I'll do #110 in a few minutes. A lot of them are very difficult to do since they require a very specific device. As much I've enjoyed helping out with this library here and there for the past few years, I'm not going to buy a random air conditioner just to try it out :)

**Comment by z3t0** [comment]
@crash7 that would mean reopening the closed PRs correct?
@joshuajnoble totally :smile:

**Comment by crash7** [comment]
I can take care of the Aiwa one (as i was the owner of the old PR).
We should split this in:
- Sharp decoding
- Aiwa decoding
- Readme & Contrib updates

@joshuajnoble exactly, if we don't find any suitable tester, we don't have a way to know if it works. We have to decide between merging the PRs or keeping the library stable being 100% sure that all the features work.

@z3t0 i'm not sure if you can push commits to submitted PRs since the branch is in another repo. If you can, then you can update them. If not, recreating them is the only way.

**Comment by z3t0** [comment]
@crash7 why not make two types of releases? We could try and verify the untested ones as much as possible and then release them as betas? Hmm, no I can't push to the other repos. We can just recreate them and reference the old ones to the new ones. I can take readme and contribution. I'll also get started on all those issues marked as "Add to Wiki". I am trying to get a sort of series of tutorials to put up on the GitHub Page.

**Comment by crash7** [comment]
@z3t0 maybe, the important thing is don't merging PRs that we aren't sure that work fine because we don't have an specific device or because they introduce breaking changes.

**Comment by z3t0** [comment]
@crash7 sure.

**Comment by z3t0** [comment]
@crash7 @joshuajnoble Please keep in mind #109 when creating new PRs

**Comment by z3t0** [comment]
@crash7 @joshuajnoble @csBlueChip I plan on merging this into the master branch soon, once I finish writing up documentation and merging the PRs I marked as WIP. The best way to find bugs is to just release it as most of us do not have the equipment needed to test all protocols.

Any thoughts?

**Comment by z3t0** [comment]
Merging as it is stable enough for the masses. Development for 2.0 will continue and be released as an update.

**Comment by z3t0** [comment]
@crom35 Please create a new issue as I am closing the experimental branch as the changes have been merged into master

**Pull request merged**